packages/core/src/utils/userop.ts

```diff
@@ -1,32 +1,104 @@
+import type { Address } from "abitype";
+import {
+  concat,
+  encodeAbiParameters,
+  hexToBigInt,
+  isAddress,
+  keccak256,
+  pad,
+  toHex,
+  type Hash,
+  type Hex,
+} from "viem";
+import type { EntryPointVersion } from "../entrypoint/types";
+import { UnexpectedEntryPointVersionError } from "../errors/entrypoint.js";
 import type {
   BigNumberish,
   Multiplier,
   UserOperationFeeOptionsField,
   UserOperationRequest,
+  UserOperationRequest_v7,
   UserOperationStruct,
+  UserOperationStruct_v6,
+  UserOperationStruct_v7,
 } from "../types";
 import { bigIntClamp, bigIntMultiply } from "./bigint.js";
-import { isBigNumberish } from "./index.js";
+import { allEqual, isBigNumberish } from "./index.js";
 
 /**
- * Utility method for asserting a {@link UserOperationStruct} is a {@link UserOperationRequest}
+ * Utility method for asserting a {@link UserOperationStruct} has valid fields for the given entry point version
  *
  * @param request a {@link UserOperationStruct} to validate
- * @returns a type guard that asserts the {@link UserOperationStruct} is a {@link UserOperationRequest}
+ * @param entryPointVersion a {@link EntryPointVersion} entry point version
+ * @returns a type guard that asserts the {@link UserOperationRequest} is valid
  */
-export function isValidRequest(
-  request: UserOperationStruct
-): request is UserOperationRequest {
+export function isValidRequest<
```

Comment:
we talked about not using `zod` in a lot of places... but parsing UserOperationStruct for validity is actually a good usecase.

if the the `EntryPointDef` had a method called `validateUserOperation` we could actually have the `EntryPoint` enforce the types. Then methods like `sendUserOperation` wouldn't necessarily need to be generic on the version and instead would just accept the union type of `UserOperationStructs`. Then the `account.entrypoint` would validate it supports the UO struct passed into it